### PR TITLE
Improve type.Table. Validate types. Remove semantic checks from parser. Groom code.

### PR DESCRIPTION
**Pull request by Renelvon**
### Parse:
- Purge all semantic checks.
- Reuse `simple_expr` rules in patterns.
### Type:
- Rewrite of `Table`:
  - Don't create new `Type` objects; reuse AST nodes.
  - Store type info using only dictionaries; leverage `__eq__` and `__hash__` of the respective AST nodes. Use `smartdicts` for retrieval of keys.
  - Allow cross-referencing of user-defined types and constructors to ease type inference.
  - Split type processing into functions for readability.
  - Propagate errors via exceptions.
- Add type validation in separate class.
- Fix bad line info in error message.
- Beautify code.
### Type test:
- Add complete suite of unit tests for `type` module.
- Run tests with parser support.
- DRY using `parserDB` and the new API.
- Enable tests due on this merge.
- Replace `unittest` primitives with `sure` primitives.
### misc:
- Fix some `pylint` warnings.

### CODE REVIEW

**Comment by dionyziz** [comment]
Please merge from master.

File: compiler/parse.py [hunk 1]
```diff
@@ -447,6 +426,11 @@ def p_miconst_simple_pattern(self, p):
         p[0] = ast.ConstExpression(ast.Int(), -p[2])
         _track(p)
 
+    def p_piconst_simple_pattern(self, p):
+        """piconst_simple_pattern : PLUS ICONST"""
+        p[0] = ast.ConstExpression(ast.Int(), p[2])
+        _track(p)
+
```
Comment (dionyziz) [inline compiler/parse.py:433]: These should be tested before refactoring. Please add some tests for float patterns with signs.

File: compiler/type.py [hunk 2]
```diff
@@ -20,24 +123,85 @@ class Table:
...
+            self._signal_error(
+                "%d:%d: error: Redefining constructor '%s'"
+                "\tPrevious definition: %d:%d",
+                constructor.lineno,
+                constructor.lexpos,
+                constructor.name,
+                existingConstructor.lineno,
+                existingConstructor.lexpos
+            )
+
```
Comment (dionyziz) [inline compiler/type.py:204]: This separation into functions is much better than what we used to have :+1:

File: compiler/type.py [hunk 3]
```diff
@@ -20,24 +123,85 @@ class Table:
...
+            )
+        else:
+            self._signal_error(
+                "%d:%d: error: Redefining user-defined type '%s'"
+                "\tPrevious definition: %d:%d",
+                newType.lineno,
+                newType.lexpos,
+                newType.name,
+                existingType.lineno,
+                existingType.lexpos
```
Comment (dionyziz) [inline compiler/type.py:176]: Good use of `smartdict` :+1:

File: tests/test_ast.py [hunk 4]
```diff
@@ -19,3 +20,58 @@ def test_regression_attr_equality(self):
 
         node_eq = ast.Node.__eq__
         node_eq.when.called_with(tdef1, tdef2).shouldnt.throw(AttributeError)
+
+    def test_builtin_type_equality(self):
+        for typecon in ast.builtin_types_map.values():
+            (typecon()).should.equal(typecon())
+
+        for typecon1, typecon2 in itertools.combinations(
+                ast.builtin_types_map.values(), 2
```
Comment (dionyziz) [inline tests/test_ast.py]: nit: single indent

File: tests/test_ast.py [hunk 5]
```diff
@@ -19,3 +20,58 @@ def test_regression_attr_equality(self):
...
+        ):
+            (typecon1()).shouldnt.equal(typecon2())
```
Comment (dionyziz) [inline tests/test_ast.py]: nit: unnecessary `()`

File: tests/test_ast.py [hunk 6]
```diff
@@ -19,3 +20,58 @@ def test_regression_attr_equality(self):
...
+
+    def test_builtin_type_set(self):
+        typeset = {typecon() for typecon in ast.builtin_types_map.values()}
+        for typecon in ast.builtin_types_map.values():
+            (typeset).should.contain(typecon())
```
Comment (dionyziz) [inline tests/test_ast.py]: nit: unnecessary `()` (and similarly in the rest of the file)

File: tests/test_type.py [hunk 7]
```diff
@@ -124,7 +65,87 @@ def test_type_process(self):
...
+            "array of (foo ref)",
+            "array [*, *] of int",
+
+            "int -> int",
+            "foo -> int",
+            "int -> foo",
+            "int ref -> int",
+            "int -> (int ref)",
+            "(array of int) -> int"
+            "int -> (array of int -> int)"
```
Comment (dionyziz) [inline tests/test_type.py]: We should add a test for functions taking functions as arguments:

`(int -> int) -> int`

File: tests/test_ast.py [hunk 8]
```diff
@@ -19,3 +20,58 @@ def test_regression_attr_equality(self):
...
+    def test_builtin_type_equality(self):
+        for typecon in ast.builtin_types_map.values():
+            (typecon()).should.equal(typecon())
+
+        for typecon1, typecon2 in itertools.combinations(
+                ast.builtin_types_map.values(), 2
+        ):
+            (typecon1()).shouldnt.equal(typecon2())
+
+    def test_builtin_type_set(self):
```
Comment (dionyziz) [inline tests/test_ast.py:31]: Why are we testing for set inclusion for builtin types, but not for user types?

**Comment by dionyziz** [comment]
This PR looks great, looking forward to merging it.

File: compiler/parse.py (hunk 1 above)
Comment (Renelvon) [inline compiler/parse.py:433]: Done.

File: tests/test_ast.py (hunk 4 above)
Comment (Renelvon) [inline tests/test_ast.py]: Fixed.

File: tests/test_type.py (hunk 7 above)
Comment (Renelvon) [inline tests/test_type.py]: Done.

File: tests/test_ast.py (hunk 8 above)
Comment (Renelvon) [inline tests/test_ast.py:31]: Tests added.

File: tests/test_ast.py (hunk 5 above)
Comment (Renelvon) [inline tests/test_ast.py]: Fixed.

File: tests/test_ast.py (hunk 6 above)
Comment (Renelvon) [inline tests/test_ast.py]: Fixed.

File: tests/test_parser.py [hunk 9]
```diff
@@ -191,10 +191,35 @@ def test_for_expr(self):
         )
 
     def test_pattern(self):
-        self._parse("true", "pattern").should.equal(self.true)
         self._parse("Red true", "pattern").should.equal(
             ast.Pattern("Red", [self.true])
         )
+        self._parse("(true)", "pattern").should.equal(self.true)
+
+        self._parse('foo', "pattern").should.equal(ast.GenidPattern("foo"))
```
Comment (dionyziz) [inline tests/test_parser.py]: nit: be consistent with your quotes

Comment (Renelvon) [inline tests/test_parser.py]: Fixed.

**Pull request merged**